lib/elixir/lib/list.ex

```diff
@@ -246,11 +248,18 @@ defmodule List do
       iex> List.last([1, 2, 3])
       3
 
+      iex> List.last([1, 2, 3 | 4])
+      4
   """
+  @spec last([]) :: nil
   @spec last([elem]) :: nil | elem when elem: var
-  def last([]), do: nil
-  def last([head]), do: head
-  def last([_ | tail]), do: last(tail)
+  @spec last(nonempty_improper_list(any, tail)) :: tail when tail: var
+  def last(list) when is_list(list), do: last_guarded(list)
```

Comment:
Since we're using brackets in all clauses except one, do we really need this additional function with the guard? Couldn't we add `is_list` to the last clause and achieve the same?

Reply:
the last clause is not a list, is an improper list tail. If we don't check with `is_list` in the main function, we cannot tell whether the user passed a non-list term from the tail of an improper list.

Reply:
Oh I see what you mean. Then this maybe:

```elixir
def last([]), do: nil
def last([head]), do: head
def last([_ | tail]) when is_list(tail), do: last(tail)
def last([_ | improper_list_tail]), do: improper_list_tail
```

? Not sure this would work but it would save us the additional function.

Reply:
you will run `is_list/0` the length of the list minus 2.
with my implementation it is run once.